sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWriteUnshardedRecords.java

```diff
@@ -758,7 +800,24 @@ public void outputWithTimestamp(
                 BigQueryStorageApiInsertError output, org.joda.time.Instant timestamp) {
               context.output(failedRowsTag, output, timestamp, GlobalWindow.INSTANCE);
             }
-          });
+          };
+      @Nullable OutputReceiver<TableRow> successfulRowsReceiver = null;
+      if (successfulRowsTag != null) {
+        successfulRowsReceiver =
+            new OutputReceiver<TableRow>() {
```

Comment:
It looks like the original windows are not preserved here. The assumption is that most of the time the customer will use FixedWindows in streaming pipelines.

Reply:
In this case it doesn't really make sense (nor is it possible) to preserve "original" windowing, since this is not really the original PCollection. In fact, even the elements aren't the same: the original PCollection may not contain TableRow elements (e.g. it might be Beam Rows) and even if it does, do the conversions back and forth it's possible that field values are slightly different.

The simples semantic here is for this PCollection to always be in the global window.

Reply:
That's what I thought and I understand the complexity. But it won't help answer "are all the incoming rows in this window processed?".  It will be non-trivial to know how a unit of work (either defined as rows in a window or identified by some id) is completed. But perhaps for some use cases having the complete set of records is needed.

Reply:
hmm, the key point here is that the output elements need to have the same timestamps as the input elements (because then you can just reapply the Window function). This is a bit tricky for the exactly-once sink, as the GroupIntoBatches loses the individual element timestamps (you just get a single timestamp for the batch element).

I'm trying to think of a way to make this work without adding a lot of extra cost or complexity.